Add documentation for transformer collocation with runtime (#464)
kserve/website: docs/modelserving/v1beta1/transformer/collocation/README.md

2. Your cluster's Istio Ingress gateway must be [network accessible](https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-control/).
3. You can find the [code samples](https://github.com/kserve/kserve/tree/master/docs/samples/v1beta1/transformer/collocation) on kserve repository.

> ## Collocation with custom container

Reviewer: ```suggestion
## Collocation with Custom Containers
```

### Deploy the InferenceService

Since, the predictor and the transformer are in the same pod, they need to listen on different ports to avoid conflict. `Transformer` is configured to listen on port 8080 (REST) and 8081 (GRPC) 
while, `Predictor` listens on port 8085 (REST). `Transformer` calls `Predictor` on port 8085 via local socket. 
Deploy the `Inferenceservice` using the below command.

> Note that, readiness probe is specified in the transformer container. This due to the limitation of Knative. You can provide `--enable_predictor_health_check` argument to allow the transformer container to check the predictor health as well. This will make sure that both the containers are healthy before the isvc is marked as ready.

Reviewer 2: ```suggestion
Note that, readiness probe is specified in the transformer container. This is due to the limitation of Knative. You can provide `--enable_predictor_health_check` argument to allow the transformer container to check the predictor health as well. This will make sure that both the containers are healthy before the isvc is marked as ready.
```

Reviewer: it is not a limitation of Knative, knative adds the default readiness TCP check for the entrypoint container, but ye if you want a proper HTTP check then you need to add it explicitly.

> Note that, readiness probe is specified in the transformer container. This is due to the limitation of Knative. You can provide `--enable_predictor_health_check` argument to allow the transformer container to check the predictor health as well. This will make sure that both the containers are healthy before the isvc is marked as ready.

Reviewer: ```suggestion
Note that, HTTP readiness probe can be specified in the transformer container to override the default TCP readiness probe. You can provide `--enable_predictor_health_check` argument to allow the transformer container to check the predictor health as well. This will make sure that both the containers are healthy before the isvc is marked as ready.
```

        image: kserve/image-transformer:latest
        args:
          - --model_name=mnist

>           - --predictor_protocol=v1

Reviewer 2: May we keep the comment or use v2?

          - --http_port=8080
          - --grpc_port=8081
          - --predictor_host=localhost:8085      # predictor listening port
          - --enable_predictor_health_check
        ports:
          - containerPort: 8080
            protocol: TCP

>         readinessProbe:

Reviewer 2: What about adding liveness too?

    predictor. The storage uri should be only present in this container. If it is specified in the transformer 
    container the isvc creation will fail.

!!! Note
    In Serverless mode, Specifying ports for predictor will result in isvc creation failure as specifying multiple ports 
    is not supported by knative. Due to this limitation predictor cannot be exposed to the outside cluster. 
    For more info see, [knative discussion on multiple ports](https://github.com/knative/serving/issues/8471).

!!! Tip

>     Check the [Transformer documentation](../torchserve_image_transformer/#transformer-specific-commandline-arguments) for list of arguments that can be passed to the transformer container.

Reviewer: We will need to start removing torch serve from the example as the project has limited maintenances.

    * Connection #0 to host localhost left intact
    {"predictions":[2]}
    ```

> ## Collocation with Runtime

Reviewer: ```suggestion
## Collocation with Model Serving Runtime
```

### Deploy the InferenceService

> Since, the predictor and the transformer are in the same pod, they need to listen on different ports to avoid conflict. `Transformer` is configured to listen on port 8080 (REST) and 8081 (GRPC) 

Reviewer: ```suggestion
Since the predictor and the transformer are in the same pod, they need to listen on different ports to avoid conflict. In this example`Transformer` is configured to listen on port 8080 (REST) and 8081 (GRPC) 
```

> while, `Predictor` listens on port 8085 (REST). `Transformer` calls `Predictor` on port 8085 via local socket. 

Reviewer 2: Maybe also mention which is the default port for gRPC for predictors.

Reviewer: it differs for different serving runtimes, can be documented on serving runtime page.

> while, `Predictor` listens on port 8085 (REST). `Transformer` calls `Predictor` on port 8085 via local socket. 

Reviewer: ```suggestion
while `Predictor` listens on port 8085 (REST) and `Transformer` calls `Predictor` on port 8085 via local socket. Please review the default listening ports for each model serving runtimes to configure properly.
```

Deploy the `Inferenceservice` using the following command:

> Note that, readiness probe is specified in the transformer container. This due to the limitation of Knative. You can provide `--enable_predictor_health_check` argument to allow the transformer container to check the predictor health as well. This will make sure that both the containers are healthy before the isvc is marked as ready.

Reviewer: ```suggestion
Note that, HTTP readiness probe is specified in the transformer container to override the default TCP readiness probe. You can provide `--enable_predictor_health_check` argument to allow the transformer container to check the predictor health as well. This will make sure that both the containers are healthy before the isvc is marked as ready.
```

        resources:
          requests:
            cpu: 100m
            memory: 256Mi
          limits:
            cpu: 1
            memory: 1Gi
EOF
```

!!! success "Expected output"
    ```{ .bash .no-copy }
    $ inferenceservice.serving.kserve.io/transformer-collocation created
    ```

### Check InferenceService status
```bash
kubectl get isvc custom-transformer-collocation
```
!!! success "Expected output"
    ```{ .bash .no-copy }
    NAME                             URL                                                         READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                              AGE
    transformer-collocation   http://transformer-collocation.default.example.com   True           100                              transformer-collocation-predictor-00001   133m
    ```

!!! Note
    If your DNS contains `svc.cluster.local`, then `Inferenceservice` is not exposed through Ingress. You need to [configure DNS](https://knative.dev/docs/install/yaml-install/serving/install-serving-with-yaml/#configure-dns) 
    or [use a custom domain](https://knative.dev/docs/serving/using-a-custom-domain/) in order to expose the `isvc`.

> ### Run a prediction

Reviewer: ```suggestion
### Run a Prediction
```

    < server: istio-envoy
    < x-envoy-upstream-service-time: 315
    < 
    * Connection #0 to host localhost left intact
    {"predictions":[2]}
    ```

## Defining Collocation In ServingRuntime

You can also define the collocation in the `ServingRuntime` and use it in the `InferenceService`. This is useful when you want to use the same transformer for multiple models.

### Create ServingRuntime

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: serving.kserve.io/v1alpha1
kind: ServingRuntime
metadata:
  name: pytorch-collocation
spec:
  annotations:
    prometheus.kserve.io/port: "8080"
    prometheus.kserve.io/path: "/metrics"
  supportedModelFormats:
    - name: pytorch
      version: "1"
      autoSelect: true
      priority: 1
  protocolVersions:
    - v1
  containers:
    - name: kserve-container
      image: pytorch/torchserve:0.9.0-cpu
      args:
        - torchserve
        - --start
        - --model-store=/mnt/models/model-store
        - --ts-config=/mnt/models/config/config.properties
      env:
        - name: "TS_SERVICE_ENVELOPE"
          value: "{% raw %}{{.Labels.serviceEnvelope}}{% endraw %}"
      securityContext:

>         runAsUser: 1000    # User ID is not defined in the Dockerfile, so we need to set it here to run as non-root

Reviewer 2: I think k8s adds it by default, ideally, we need to run a random uid, no?

          drop:
            - ALL
      resources:
        requests:
          cpu: "1"
          memory: 2Gi
        limits:
          cpu: "1"
          memory: 2Gi
    
    - name: transformer-container    # Do not change the container name
      image: kserve/image-transformer:latest
      args:
        - --model_name={% raw %}{{.Labels.modelName}}{% endraw %}
        - --predictor_protocol=v1
        - --http_port=8080
        - --grpc_port=8081
        - --predictor_host=localhost:8085      # predictor listening port
        - --enable_predictor_health_check      # transformer checks for predictor health before marking itself as ready
      ports:
        - containerPort: 8080
          protocol: TCP
      readinessProbe:
          httpGet:
            path: /v1/models/{% raw %}{{.Labels.modelName}}{% endraw %}
            port: 8080
          initialDelaySeconds: 5
          periodSeconds: 10
      resources:
        requests:
          cpu: 100m
          memory: 256Mi
        limits:
          cpu: 1
          memory: 1Gi
EOF
```

!!! note
    Do not specify ports for predictor in the serving runtime for Serverless deployment. This is not supported by knative. 

>     For more info see, [knative discussion on multiple ports](https://github.com/knative/serving/issues/8471).

Reviewer 2: ```suggestion
    For more information, please take a look at [knative discussion on multiple ports](https://github.com/knative/serving/issues/8471).
```
